lib/pbench/server/cache_manager.py

```diff
@@ -93,39 +95,57 @@ def __str__(self) -> str:
         return f"An error occurred while changing file permissions of {self.tarball}: {self.error}"
 
 
-class FileInfo:
-    def __init__(self, dir_path: Path, path: Path):
-        """Collects the file info
+class CacheType(Enum):
 
-        Args:
-            dir_path: root directory parent path
-            path: path to a file/directory
-        """
+    FILE = auto()
+    DIRECTORY = auto()
+    SYMLINK = auto()
+    OTHER = auto()
+
+
+@dataclass
+class CacheObject:
+
+    name: str = None
+    location: Path = None
```

Comment:
> I have the impression that this isn't the best use of `@dataclass`: I think you may be trying to do too much with it. I recommend writing a function (like a "factory") which takes your two inputs (i.e., `dir_path` and `path`) and which constructs and returns a `CacheObject` in a fairly straightforward way.

but if this is not what `__post_init__` method to be used for then what? I know we are computing a lot of things here but at the end it is returning a dataclass object and the solution that you are suggesting of using a factory function and returning a dataclass , it would be doing the same thing.

Reply:
First off, I freely concede that the code that you are proposing works, that you've found and engaged the special support for extending `@dataclass` construction, and that we _could_ go with what you have.  Second, while we could explore what `__post_init__()` is supposed to be used for, I would suggest that we not go there, because, if we adopt a more straightforward approach, then we don't need to consider magic like `__post_init__()`.

The problem that we're trying to solve is that we want to build some data structures containing multiple fields so that we can exchange collections of heterogenous values between functions.  We have lists, tuples, and dictionaries, but they are all weakly specified and their contained elements are untyped.  What we should probably have reached for are [`NamedTuple`](https://docs.python.org/3/library/typing.html#typing.NamedTuple)'s:  they allow us to specify the data members, specify their types, and reference them by name.  But, `NamedTuple`'s have some drawbacks (IIRC, you cannot create subclasses of them), but they would probably suffice for our present purpose.

The alternative to using a `NamedTuple` is to create a full-blown `class`.  But, this is somewhat cumbersome if all we want is a rigid container to stuff our favorite values into:  specifically, if we rolled our own class, we would need to define an `__init__()` method which takes all the values that we want to put into it, and which initializes all the fields to those values...which is boring and takes up a bunch of space.

To the rescue comes `@dataclass`!  By wrapping our class definition with that decorator, the annoying `__init__()` function gets generated for us implicitly, and we can use the class c'tor even though it doesn't appear in our code.  (And, we get a bunch of other bonus functions generated for us, like `__str__()`, `__lt__()`, and `__eq__()` so we can print, compare, and sort instances of our class.)

If we take some care in how we define our class we can arrange default values for some or all of the data members, so we don't have to specify them in all the calls to the c'tor.  This can allow the more straightforward constructions to be done with more straightforward code.

The problem with this that you're wrestling with is that we want to our instances to have three, four, or six members whose values are set based on two input parameters.  So, you're thinking that the c'tor should take just two parameters.  But, the whole point of `@dataclass` is to make it easy to initialize all the members.  You discovered a mechanism to allow you to pass in additional parameters beyond the ones used to initialize the members, but, really, you would rather _not_ pass in values to initialize the members, so you set them up with default values of `None` (but, then that means that you have to set their type hints to `Optional`...), and then you rely on the `__post_init__()` method to actually initialize the members, and now we've strayed really far from the original idea of just handing the values to the generated `__init__()` function.

So, to get us back to the design center, what we need is a function which accepts our two inputs, generates the 3/4/6 values, instantiates an object using a straightforward construction, and returns it.  That is, instead of having the caller call the c'tor for a rather mangled `@dataclass`, the caller calls a different function which wraps a straightforward `@dataclass`.

The other advantage of the factory approach is that it can provide a polymorphic return.  That is, with your current approach, we only have one kind of `CacheObject`, and, therefore, it has to contain all six fields, even though we don't use half of them most of the time.  If we use a factory, it can return a subclass for the cases where we need additional fields.  (And, because we're using `@dataclass` instead of `NamedTuple`, we can implement subclasses straightforwardly.)

So, we can do something like this (which I think Dave suggested originally):
```python
@dataclass
class CacheObject:
    name: str = None
    location: Path = None
    type: CacheType = None

@dataclass
class CacheFile(CacheObject):
    size: int = None

@dataclass
class CacheSymlink(CacheObject):
    resolve_path: Path = None
    resolve_type: str = None

def make_cache_object(dir_path: Path, path: Path):
        name = path.name
        location = path.relative_to(dir_path)
        if path.is_symlink():
            try:
                link_path = path.readlink()
                if not link_path.is_absolute():
                    raise ValueError("symlink path is absolute")
                resolve_path = path.resolve(strict=True)
                rpath = resolve_path.relative_to(dir_path)
            except (FileNotFoundError, ValueError):
                rpath = link_path
                rtype = CacheType.SYMLINK
            else:
                if resolve_path.is_dir():
                    rtype = CacheType.DIRECTORY
                elif resolve_path.is_file():
                    rtype = CacheType.FILE
                else:
                    rtype = CacheType.OTHER
            return CacheSymlink(name, location, CacheType.SYMLINK, rpath, rtype)
        elif path.is_file():
            return CacheFile(name, location, CacheType.FILE, path.stat().st_size)
        elif path.is_dir():
            return CacheObject(name, location, CacheType.DIRECTORY)
        else:
            return CacheObject(name, location, CacheType.OTHER)
```
The caller can then do the following:
```
    co = make_cache_object(dir_path, path)
    print(f"My name is {co.name!r}, I'm at {co.location}, with type {co.type}")
    if co.type is CacheType.FILE:
        print("I have a size:", co.size)
    elif co.type is CacheType.SYMLINK:
        print(f"I refer to {co.resolve_path} which is a {co.resolve_type}")
    print(f"And, my cache object looks like {repr(co)}")
```

Reply:
I've played with this a little bit since I first posted it, and I've made two revisions:
- I removed the default values from the `@dataclass`'s, which I'd copied from your code but shouldn't have kept;
- I noted that the `is_absolute()` test was _backwards_, so I corrected it above.

Also, I've convinced myself that the assignment of `rtype = CacheType.SYMLINK` should be `rtype = CacheType.OTHER`:  in the case where the referent is not found, obviously we don't know what kind it is, so, now that we have the option, `OTHER` makes more sense than `SYMLINK`; similarly, if the referent is outside the results area, then it's referring to a file on the original host (not where we've unpacked the result), so we wouldn't know what kind of file it is; and, if the link is an absolute path, it's reasonable for us to assume that the file is inaccessible to us, and so, again, `OTHER` is a better choice.  (The only time that `SYMLINK` would be the best choice would be if the current file referred to another symlink in the results area which itself referred to something unknown, and I'm willing to say that the current file should inherit the characteristic of the target, which means that the result would be `OTHER`.)

Reply:
Thank you @webbnh for the code and elaborate explaination. Everything seems to be good but there is one issue this is not returning a consistent objenct. If it is a Directory then there is no resolbe_path or resolve_type and same goes for file. So, now do we want to move with different object types and test them differently? like a different test for directory, separate for files and same for symlinks.

Reply:
The answer to your question is tied up in how we intend to use these objects.

For most uses, treating the factory's return as a `CacheObject` (i.e., the base class) will be sufficient and unambiguous.  That is, in most cases, I think all we care about is the location and name of the cache entry.  It will only be in specialized circumstances when we want to know, e.g., the file size or the symlink resolutions, and I expect that we will generally already know which type of object we're supposed to have when we want to reference the subclass fields.  (Admittedly, we will potentially find instances of `FILE`, `SYMLINK`, and `OTHER` in the same bucket, but we can use `CacheObject.type` to figure out what we've got...which is probably what we would do if we stuck to just a single `CacheOjbect` class.  Interestingly, if we go with the class hierarchy, we can dispense with the `type` field, because we can always ask what the type is of the object that we have (e.g., using `type(obj) is CacheSymlink` or `isinstanceof(obj, CacheSymlink)`.)

_[I realize that earlier I counseled against omitting fields from some types' instances (and I continue to stand by that request 🙂), and I concede that the difference between that situation and this one is very fine.  But, the principal of the thing is that, here, we have distinct types (`CacheObject`, `CacheSymlink`, et al.), and so there is no reason to expect that there would be a `size` in a `CacheObject` which was not a `CacheFile`; whereas what we had before was a generic dictionary which might describe any of the types, and therefore it is less surprising if the unused keys are present all the time, because, in effect the "type" of the dictionary is the same for all the cases.  (One could certainly claim that in fact, the "type" of the dictionary was _different_ for each of the cases, and that would justify omitting the fields from the "types" which don't need them, but my problem with that is that in all the cases, it's still a `dict` and one has to consult the `type` key to figure out what one has; whereas, with the class hierarchy, the "type" is baked into the definition, and so there's no surprise that certain fields are or are not present in a given instance.)]_

Reply:
So...now this discussion needs to progress from how we represent and handle the cache map entries (which `traverse_cmap()` returns) on to what `get_info()` returns.  😁 

As Dave has been saying, having `get_info()` return a `dict`/JSON-blob is adequate, but we could do better.

And, I don't really like the alternating `dict`'s which represent the cache map (i.e., at one level, there's a `dict` with keys `"details"` and `"children"`; at the next level there's a `dict` whose keys are filenames; and then the next level has keys `"details"` and `"children"` again).

If we go with this `CacheObject` hierarchy, then it would be easy to add a `CacheDirectory` subclass which includes a data member for the `"children"` `dict` with the filename keys and the `CacheObject` values.  This would remove the alternation which I dislike (since we could descend directly from one `CacheDirectory` instance to the next `CacheObject`), and it would enable `get_info()` to simply return the `CacheObject` produced by `traverse_cmap()`, which I think would make Dave happy.  (The `CacheDirectory` subclass might have to maintain files and subdirectories in separate lists which is what `get_info()` is doing now -- I don't know what our requirements are, here, since I don't know who is going to be calling `get_info()` -- but `get_info()` would no longer have to do an object conversion to produce its return value (in fact, `Tarball.get_info()` might go away altogether, because `CacheManager.get_info()` could call `Tarball.traverse_cmap()` directly and just return what it returns).

The biggest question is, should we do this conversion in _this_ PR, or can we declare victory here (soon...) and make that change in the next PR?

Reply:
I updated the return type when Dave suggested but forgot to update that in his comment.

I have added a documentation referring the get_info return type and added the symlink test_cases as well.